text/0000-promise.md

```diff
@@ -0,0 +1,460 @@
+
+The `JsPromise` API consists of two structs:
+
+* `JsPromise`. Opaque value; only useful for passing back to JavaScript.
+* `Deferred`. Handle for resolving and rejecting the related `Promise`.
+
+They are most idiomatically crated by `cx.promise()`, but construction is mirrored across `JsPromise` and `Deferred`.
+
+```rust
+trait Context {
+    fn promise(&mut self) -> (Deferred, Handle<JsPromise>);
+}
+
+impl JsPromise {
+    pub fn new<'a, C: Context<'a>>(cx: &mut C) -> (Deferred, Handle<'a, JsPromise>) {
+        todo!()
+    }
+}
+
+impl Deferred {
+    pub fn new<'a, C: Context<'a>>(cx: &mut C) -> (Deferred, Handle<'a, JsPromise>) {
+        todo!()
+    }
+}
+```
+
+#### `JsPromise`
+
+Opaque handle that represents a JavaScript `Promise`. It is not threadsafe.
+
+```rust
+/// A JavaScript `Promise`
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub struct JsPromise(raw::Local);
+
+impl Value for JsPromise {}
+
+impl Managed for JsPromise {
+    fn to_raw(self) -> raw::Local { self.0 }
+    fn from_raw(h: raw::Local) -> Self { JsPromise(h) }
+}
+
+impl ValueInternal for JsPromise {
+    fn name() -> String { "promise".to_string() }
+    fn is_typeof<Other: Value>(env: Env, other: Other) -> bool {
+        unsafe { neon_runtime::tag::is_promise(env.to_raw(), other.to_raw()) }
+    }
+}
+
+impl Object for JsPromise {}
+
+impl JsPromise {
+    pub fn new<'a, C: Context<'a>>(cx: &mut C) -> (Deferred, Handle<'a, JsPromise>) {
+        todo!()
+    }
+}
+```
+
+#### `Deferred`
+
+`Send` handle for resolving or rejecting a `JsPromise`.
+
+Resolving a `Deferred` always takes ownership of `self` and consumes the `Deferred` to prevent double settlement.
+
+```rust
+pub struct Deferred(*mut c_void);
+
+unsafe impl Send for Deferred {}
+
+impl Deferred {
+    /// Resolves a deferred Promise with the value
+    pub fn resolve<'a, T: Value, C: Context<'a>>(self, cx: &mut C, v: Handle<T>);
+
+    /// Rejects a deferred Promise with the error
+    pub fn reject<'a, E: Value, C: Context<'a>>(self, cx: &mut C, err: Handle<E>);
+
+    /// Resolves or rejects a Promise with the result of a closure
+    pub fn settle_with<'a, T, C, F>(self, cx: &mut C, f: F)
+    where
+        T: Value,
+        C: Context<'a>,
+        F: FnOnce(&mut C) -> JsResult<'a, T>;
+}
+```
+
+Similar to [`Root`](https://github.com/neon-bindings/rfcs/pull/32), if a `Deferred` is not resolved or rejected, the `Promise` chain will leak. To help catch this and guide users towards a correct implementation, `Deferred` should `panic` on `Drop` if not used on Node-API < 4 and be dropped on a global drop queue in Node-API >= 4.
+
+```rust
+impl std::ops::Drop for Deferred {
+    fn drop(&mut self) {
+        panic!("JsPromise leaked. Deferred must be used.");
+    }
+}
+```
+
+#### `Channel` extension
+
+A new method `Channel::send_and_settle` is added to reduce the boilerplate of nested closures from `Channel::send` and `Deferred::settle_with`. Example without the convenience method:
```

Comment:
I'd propose we call this `Channel::settle_with`, for a few reasons:
- It's a little more ergonomic / less of a mouthful than `send_and_settle`
- It hints at being a shorthand by having the same name as `Deferred::settle_with`
- It's nice for refactoring that you can change from sync to async with a slightly easier-to-understand change